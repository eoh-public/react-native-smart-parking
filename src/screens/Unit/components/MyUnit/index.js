import React, { useMemo, useCallback } from 'react';
import {
  View,
  Image,
  TouchableOpacity,
  StyleSheet,
  Dimensions,
} from 'react-native';
import Carousel from 'react-native-snap-carousel';
import { useNavigation } from '@react-navigation/native';
import t from 'i18n';

import { Colors, Images } from '../../../../configs';
import Text from '../../../../commons/Text';
import MyUnitDevice from '../MyUnitDevice';
import Routes from '../../../../utils/Route';
import { colorOpacity } from '../../../../utils/Converter/color';

let screenWidth = Dimensions.get('window').width;

const MyUnit = ({ myUnits }) => {
  const navigation = useNavigation();
  const carouselItems = useMemo(() => myUnits, [myUnits]);
  const goToDetail = useCallback(
    (item) => {
      navigation.navigate(Routes.UnitStack, {
        screen: Routes.UnitDetail,
        params: {
          unitId: item.id,
          unitData: item,
        },
      });
    },
    [navigation]
  );
  const _renderItem = useCallback(
    ({ item, index }) => {
      const paddingLeft = index === 0 ? 0 : 8;
      const paddingRight = index === carouselItems.length - 1 ? 0 : 8;
      return (
        <View
          style={{
            paddingLeft: paddingLeft,
            paddingRight: paddingRight,
          }}
        >
          <TouchableOpacity
            onPress={() => goToDetail(item)}
            style={styles.btnItem}
            activeOpacity={0.75}
          >
            <View style={styles.overlay} />
            <Image
              style={styles.bgMyUnit}
              source={{ uri: item.background }}
              defaultSource={Images.BgUnit}
              resizeMode="cover"
            />
            <Text style={styles.title}>{item.name}</Text>
          </TouchableOpacity>
          {item.abstract_sensors.map((sensor, indexSensor) => (
            <MyUnitDevice key={indexSensor} sensor={sensor} />
          ))}
        </View>
      );
    },
    [carouselItems.length, goToDetail]
  );

  return (
    <View style={styles.container}>
      {carouselItems.length ? (
        <Carousel
          layout={'default'}
          data={carouselItems}
          sliderWidth={screenWidth}
          itemWidth={screenWidth - 32}
          renderItem={_renderItem}
          inactiveSlideScale={1}
        />
      ) : (
        <View>
          <Text>{t('text_no_units')}</Text>
        </View>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.White,
    flexDirection: 'row',
    justifyContent: 'center',
    paddingBottom: 16,
  },
  overlay: {
    backgroundColor: colorOpacity(Colors.Black, 0.4),
    zIndex: 2,
    left: 0,
    top: 0,
    position: 'absolute',
    width: '100%',
    height: '100%',
    borderRadius: 10,
  },
  bgMyUnit: {
    position: 'absolute',
    left: 0,
    top: 0,
    right: 0,
    width: '100%',
    height: '100%',
    borderRadius: 10,
  },
  title: {
    position: 'absolute',
    zIndex: 3,
    fontWeight: 'bold',
    fontSize: 20,
    lineHeight: 28,
    color: Colors.White,
    bottom: 16,
    left: 16,
  },
  btnItem: {
    height: 121,
    marginBottom: 16,
  },
});

export default MyUnit;
