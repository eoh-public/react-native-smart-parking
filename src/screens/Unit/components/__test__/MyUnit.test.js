import React from 'react';
import renderer, { act } from 'react-test-renderer';

import MyUnit from '../MyUnit';

describe('Test MyUnit', () => {
  let tree;
  test('create MyUnit', () => {
    const myUnits = [
      {
        abstract_sensors: [],
        name: 'unittest 1',
      },
    ];
    act(() => {
      tree = renderer.create(<MyUnit myUnits={myUnits} />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
