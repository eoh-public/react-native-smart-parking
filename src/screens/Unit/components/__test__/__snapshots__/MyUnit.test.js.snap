// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MyUnit create MyUnit 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "paddingBottom": 16,
    }
  }
>
  <RCTScrollView
    CellRendererComponent={[Function]}
    activeSlideAlignment="center"
    activeSlideOffset={20}
    apparitionDelay={0}
    automaticallyAdjustContentInsets={false}
    autoplay={false}
    autoplayDelay={1000}
    autoplayInterval={3000}
    callbackOffsetMargin={5}
    containerCustomStyle={Object {}}
    contentContainerCustomStyle={Object {}}
    contentContainerStyle={
      Array [
        Object {
          "paddingLeft": 16,
          "paddingRight": 16,
        },
        Object {},
      ]
    }
    data={
      Array [
        Object {
          "abstract_sensors": Array [],
          "name": "unittest 1",
        },
      ]
    }
    decelerationRate="fast"
    directionalLockEnabled={true}
    disableScrollViewPanResponder={false}
    disableVirtualization={false}
    enableSnap={true}
    firstItem={0}
    forwardedRef={[Function]}
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    hasParallaxImages={false}
    horizontal={true}
    inactiveSlideOpacity={0.7}
    inactiveSlideScale={1}
    inactiveSlideShift={0}
    initialNumToRender={7}
    initialScrollIndex={0}
    invertStickyHeaders={false}
    inverted={false}
    itemWidth={718}
    keyExtractor={[Function]}
    layout="default"
    loop={false}
    loopClonesPerSide={3}
    maxToRenderPerBatch={11}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    overScrollMode="never"
    pinchGestureEnabled={false}
    pointerEvents="auto"
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEnabled={true}
    scrollEventThrottle={1}
    scrollsToTop={false}
    shouldOptimizeUpdates={true}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    slideStyle={Object {}}
    sliderWidth={750}
    snapToOffsets={
      Array [
        0,
      ]
    }
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flexDirection": "row",
        "width": 750,
      }
    }
    updateCellsBatchingPeriod={50}
    useExperimentalSnap={false}
    useScrollView={false}
    vertical={false}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={11}
  >
    <View>
      <View
        CellRendererComponent={[Function]}
        cellKey="flatlist-item-0"
        fillRateHelper={
          FillRateHelper {
            "_anyBlankStartTime": null,
            "_enabled": false,
            "_getFrameMetrics": [Function],
            "_info": Info {
              "any_blank_count": 0,
              "any_blank_ms": 0,
              "any_blank_speed_sum": 0,
              "mostly_blank_count": 0,
              "mostly_blank_ms": 0,
              "pixels_blank": 0,
              "pixels_sampled": 0,
              "pixels_scrolled": 0,
              "sample_count": 0,
              "total_time_spent": 0,
            },
            "_mostlyBlankStartTime": null,
            "_samplesStartTime": null,
          }
        }
        horizontal={true}
        index={0}
        inversionStyle={null}
        item={
          Object {
            "abstract_sensors": Array [],
            "name": "unittest 1",
          }
        }
        onUnmount={[Function]}
        onUpdateSeparators={[Function]}
        parentProps={
          Object {
            "CellRendererComponent": [Function],
            "activeSlideAlignment": "center",
            "activeSlideOffset": 20,
            "apparitionDelay": 0,
            "automaticallyAdjustContentInsets": false,
            "autoplay": false,
            "autoplayDelay": 1000,
            "autoplayInterval": 3000,
            "callbackOffsetMargin": 5,
            "collapsable": undefined,
            "containerCustomStyle": Object {},
            "contentContainerCustomStyle": Object {},
            "contentContainerStyle": Array [
              Object {
                "paddingLeft": 16,
                "paddingRight": 16,
              },
              Object {},
            ],
            "data": Array [
              Object {
                "abstract_sensors": Array [],
                "name": "unittest 1",
              },
            ],
            "decelerationRate": "fast",
            "directionalLockEnabled": true,
            "disableScrollViewPanResponder": false,
            "disableVirtualization": false,
            "enableSnap": true,
            "firstItem": 0,
            "forwardedRef": [Function],
            "getItem": [Function],
            "getItemCount": [Function],
            "getItemLayout": [Function],
            "hasParallaxImages": false,
            "horizontal": true,
            "inactiveSlideOpacity": 0.7,
            "inactiveSlideScale": 1,
            "inactiveSlideShift": 0,
            "initialNumToRender": 7,
            "initialScrollIndex": 0,
            "inverted": false,
            "itemWidth": 718,
            "keyExtractor": [Function],
            "layout": "default",
            "loop": false,
            "loopClonesPerSide": 3,
            "maxToRenderPerBatch": 11,
            "onEndReachedThreshold": 2,
            "onLayout": [Function],
            "onMomentumScrollEnd": [Function],
            "onScroll": [Function],
            "onTouchEnd": [Function],
            "onTouchStart": [Function],
            "overScrollMode": "never",
            "pinchGestureEnabled": false,
            "pointerEvents": "auto",
            "removeClippedSubviews": false,
            "renderItem": [Function],
            "scrollEnabled": true,
            "scrollEventThrottle": 1,
            "scrollsToTop": false,
            "shouldOptimizeUpdates": true,
            "showsHorizontalScrollIndicator": false,
            "showsVerticalScrollIndicator": false,
            "slideStyle": Object {},
            "sliderWidth": 750,
            "snapToOffsets": Array [
              0,
            ],
            "style": Object {
              "flexDirection": "row",
              "width": 750,
            },
            "updateCellsBatchingPeriod": 50,
            "useExperimentalSnap": false,
            "useScrollView": false,
            "vertical": false,
            "viewabilityConfigCallbackPairs": Array [],
            "windowSize": 11,
          }
        }
        style={
          Array [
            Array [
              Object {
                "flexDirection": "row",
              },
              null,
            ],
            Object {
              "zIndex": 1,
            },
          ]
        }
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "opacity": 1,
              "width": 718,
            }
          }
        >
          <View
            style={
              Object {
                "paddingLeft": 0,
                "paddingRight": 0,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "height": 121,
                  "marginBottom": 16,
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "backgroundColor": "rgba(0,0,0,0.4)",
                    "borderRadius": 10,
                    "height": "100%",
                    "left": 0,
                    "position": "absolute",
                    "top": 0,
                    "width": "100%",
                    "zIndex": 2,
                  }
                }
              />
              <Image
                defaultSource={
                  Object {
                    "testUri": "../../../App/Images/Unit/UnitBG.png",
                  }
                }
                resizeMode="cover"
                source={
                  Object {
                    "uri": undefined,
                  }
                }
                style={
                  Object {
                    "borderRadius": 10,
                    "height": "100%",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "width": "100%",
                  }
                }
              />
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000",
                      "fontFamily": "SFProDisplay-Bold",
                      "fontSize": undefined,
                      "lineHeight": undefined,
                      "textAlign": "left",
                      "textDecorationLine": "none",
                    },
                    Object {
                      "bottom": 16,
                      "color": "#FFFFFF",
                      "fontSize": 20,
                      "fontWeight": null,
                      "left": 16,
                      "lineHeight": 28,
                      "position": "absolute",
                      "zIndex": 3,
                    },
                  ]
                }
              >
                unittest 1
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
